bigframes/session/__init__.py

```diff
@@ -666,142 +639,64 @@ def _read_gbq_table(
-                assert current_timestamp is not None
-                table_expression = self.ibis_client.sql(
-                    bigframes_io.create_snapshot_sql(table_ref, current_timestamp)
+                column_values = [
+                    table_expression[col] for col in table_expression.columns
+                ]
+                array_value = core.ArrayValue.from_ibis(
+                    self,
+                    table_expression,
+                    columns=column_values,
+                    hidden_ordering_columns=[],
+                    ordering=ordering,
                 )
             else:
-                # Make sure when we generate an ordering, the row_number()
-                # coresponds to the index columns.
-                table_expression = table_expression.order_by(index_cols)
-                warnings.warn(
-                    textwrap.dedent(
-                        f"""
-                        Got a non-unique index. A consistent ordering is not
-                        guaranteed. DataFrame has {total_count} rows,
-                        but only {distinct_count} distinct index values.
-                        """,
-                    )
-                )
-
-            # When ordering by index columns, apply limit after ordering to
-            # make limit more predictable.
-            if max_results is not None:
-                table_expression = table_expression.limit(max_results)
+                array_value = self._create_total_ordering(table_expression)
         else:
-            if max_results is not None:
-                # Apply limit before generating rownums and creating temp table
-                # This makes sure the offsets are valid and limits the number of
-                # rows for which row numbers must be generated
-                table_expression = table_expression.limit(max_results)
-            table_expression, ordering = self._create_sequential_ordering(
-                table=table_expression,
-                api_name=api_name,
-            )
-            hidden_cols = (
-                (ordering.total_order_col.column_id,)
-                if ordering.total_order_col
-                else ()
-            )
-            assert len(ordering.ordering_value_columns) > 0
-            is_total_ordering = True
-            # Block constructor will generate default index if passed empty
-            index_cols = []
-            index_labels = []
-
-        return self._read_gbq_with_ordering(
-            table_expression=table_expression,
-            col_order=col_order,
-            index_cols=index_cols,
-            index_labels=index_labels,
-            hidden_cols=hidden_cols,
-            ordering=ordering,
-            is_total_ordering=is_total_ordering,
-            api_name=api_name,
+            array_value = self._create_total_ordering(table_expression)
+
+        if col_order:
+            array_value = array_value.select_columns(tuple(col_order))
```

Comment:
As we discussed in chat, we should really do the column filter before making the hash-based ordering. Any chance we can do that now? Otherwise, let's file an issue and add a TODO.

Reply:
yeah, definitely want to minimize the set of columns hashed (even if it affects the ordering). done